src/execution/window_executor.cpp

```diff
@@ -824,76 +912,57 @@ void ExclusionFilter::ResetMask(idx_t row_idx, idx_t offset) {
 	}
 }
 
-//===--------------------------------------------------------------------===//
-// WindowExecutor
-//===--------------------------------------------------------------------===//
-static void PrepareInputExpressions(const vector<unique_ptr<Expression>> &exprs, ExpressionExecutor &executor,
-                                    DataChunk &chunk) {
-	if (exprs.empty()) {
-		return;
+column_t WindowSharedExpressions::RegisterExpr(const unique_ptr<Expression> &expr, Expressions &shared) {
+	auto pexpr = expr.get();
+	if (!pexpr) {
+		return DConstants::INVALID_INDEX;
 	}
 
-	vector<LogicalType> types;
-	for (idx_t expr_idx = 0; expr_idx < exprs.size(); ++expr_idx) {
-		types.push_back(exprs[expr_idx]->return_type);
-		executor.AddExpression(*exprs[expr_idx]);
+	column_t result = shared.size();
+	for (column_t i = 0; i < shared.size(); ++i) {
+		if (pexpr->Equals(*shared[i])) {
```

Comment:
Should we check if the expression is volatile before sharing it? Currently we can share volatile expressions, leading to e.g. `random()` being the same in multiple window functions:

```sql
D SELECT list(random()) OVER (ORDER BY id), max(random()) OVER (ORDER BY id) FROM range(3) t(id);
┌───────────────────────────────────────────────────────────────┬──────────────────────────────────┐
│               list(random()) OVER (ORDER BY id)               │ max(random()) OVER (ORDER BY id) │
│                           double[]                            │              double              │
├───────────────────────────────────────────────────────────────┼──────────────────────────────────┤
│ [0.02444483428537218]                                         │              0.02444483428537218 │
│ [0.02444483428537218, 0.2667830996175638]                     │               0.2667830996175638 │
│ [0.02444483428537218, 0.2667830996175638, 0.6447218465599317] │               0.6447218465599317 │
└───────────────────────────────────────────────────────────────┴──────────────────────────────────┘

```